### PR TITLE
Import legacy BMT implementation from Swarm

### PR DESCRIPTION
**Pull request by nolash**
Minimal changes are made to the implementation in order for it to conform to the interface definition. Aim is to have a usable implementation early on. Improved implementation and/or improved clarity of code can be added later.

The PR does _not_ change any functionality

### CODE REVIEW

**Comment by nolash** [comment]
Unblocks https://github.com/ethersphere/bee/issues/39

File: legacy/bmt.go [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package bmt provides a binary merkle tree implementation used for swarm chunk hash
```
Comment (acud) [inline legacy/bmt.go]: `Package legacy`

File: legacy/bmt.go [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+	"errors"
+	"fmt"
+	"hash"
+	"strings"
+	"sync"
+	"sync/atomic"
+)
+
+/*
+Binary Merkle Tree Hash is a hash function over arbitrary datachunks of limited size.
```
Comment (acud) [inline legacy/bmt.go]: consider moving this stuff into a `doc.go` file?

File: legacy/bmt.go [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+var (
+	ZeroSpan = make([]byte, 8)
+)
+
+// BaseHasherFunc is a hash.Hash constructor function used for the base hash of the BMT.
+// implemented by Keccak256 SHA3 sha3.NewLegacyKeccak256
+type BaseHasherFunc func() hash.Hash
+
+// Hasher a reusable hasher for fixed maximum size chunks representing a BMT
+// - implements the hash.Hash interface
```
Comment (acud) [inline legacy/bmt.go]: not sure if this will render correctly in godoc

File: legacy/bmt.go [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+
+// BlockSize implements hash.Hash and file.SectionWriter
+func (h *Hasher) BlockSize() int {
+	return 2 * h.pool.SegmentSize
+}
+
+// Sum returns the BMT root hash of the buffer
+// using Sum presupposes sequential synchronous writes (io.Writer interface)
+//
+// Implements hash.Hash
```
Comment (acud) [inline legacy/bmt.go]: these comments could probably be removed and replaced by a nil interface check

File: reference/doc.go [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package bmt is a simple nonconcurrent reference implementation for hashsize segment based
```
Comment (acud) [inline reference/doc.go]: package reference

**Review by acud — APPROVED** [review]
the usual rant. otherwise looks good

File: legacy/bmt.go (hunk 4 above)
Comment (nolash) [reply legacy/bmt.go]: Shouldn't we document the fact that it implements an interface? That seems natural to me?

Comment (acud) [reply legacy/bmt.go]: a nil interface check would do; self-documenting code

**Review by acud — APPROVED** [review]

**Review by janos — APPROVED** [review]
Thanks, LGTM.

**Pull request merged**